src/stan/model/indexing/assign.hpp

```diff
@@ -14,6 +13,29 @@ namespace stan {
 
 namespace model {
 
+namespace internal {
+/**
+ * Does an NAND to require the lhs is not a var_matrix and the rhs is not an
+ *  eigen matrix.
+ */
+template <typename T1, typename T2>
+using require_nand_var_matrix_or_eigen
+    = require_not_t<stan::math::conjunction<is_var_matrix<T1>, is_eigen<T2>>>;
+/**
+ * Base case of assignment
+ * @tparam T1 Any type that's not a var matrix.
+ * @tparam T2 Any type that's not a var matrix.
+ * @param x The value to assign to
+ * @param y The value to assign from.
+ */
+template <typename T1, typename T2,
+          require_nand_var_matrix_or_eigen<T1, T2>* = nullptr>
```

Comment:
Reading nand makes my head hurt. Also I think this is wrong and should be just `require_not_var_matrix<T2>`.You can also add `require_any_t<is_var_matrix<T1>, is_eigen<T1>>` if you want, but it is not really needed.

Reply:
Yeah I did a bad job at naming this, ended up simplifying it to `require_any_not_t<is_var_matrix<T1>, is_eigen<T2>>`. It has to be the opposite of the one specialization right?